svchost/disco/disco.go

```diff
@@ -0,0 +1,197 @@
+// Package disco handles Terraform's remote service discovery protocol.
+//
+// This protocol allows mapping from a service hostname, as produced by the
+// svchost package, to a set of services supported by that host and the
+// endpoint information for each supported service.
+package disco
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"mime"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/hashicorp/terraform/svchost"
+	"github.com/hashicorp/terraform/svchost/auth"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+const (
+	discoPath        = "/.well-known/terraform.json"
+	maxRedirects     = 3               // arbitrary-but-small number to prevent runaway redirect loops
+	discoTimeout     = 4 * time.Second // arbitrary-but-small time limit to prevent UI "hangs" during discovery
+	maxDiscoDocBytes = 1 * 1024 * 1024 // 1MB - to prevent abusive services from using loads of our memory
+)
+
+var userAgent = fmt.Sprintf("Terraform/%s (service discovery)", terraform.VersionString())
+var httpTransport = http.DefaultTransport // overridden during tests, to skip TLS verification
```

Comment:
let's use `cleanhttp.DefaultPooledTransport` as the default here. I've been trying to replace usage of the `DefaultClient` and `DefaultTransport` throughout, as we've had imported packages alter those in the past.